### DISCUSSION TITLE
Preview changes (sync diff) from arbitrary branches

### DISCUSSION THREAD

**Discussion by edsonmarquezani** [argoproj/argo-cd#9056] [Q&A] [answered]
Currently, I have a CI/CD setup where I use `helm diff` to preview what was going to change if a given PR was applied. This is run in PR phase, so changes can be reviewed before being merged and actually applied. Using Github Actions, it's very easy to report the diff output as a comment. See that this doesn't change anything and all PR changes could even be discarded without never getting into main branches.

So, I was wondering if there was a way to do that with ArgoCD. Is there a way to preview a diff without actually merging the changes to the application branch? One way I see it would be to temporarily changing the application branch (let's say, from `main` to `<pr-branch>`) and getting the diff. But this is definitely not a good option.

Is there any customization feature (plugins, hook, etc) that could allow implementing this?

**Comment by rishabh625** [comment]
Didn't got you completely , but you can view diff for applications, https://argo-cd.readthedocs.io/en/stable/user-guide/commands/argocd_app_diff before syncing it if sync policy is set to manual

**Reply by edsonmarquezani** [reply]
Yes, but only once the content is actually merged. What if I want to preview changes in a PR, for example? Like: "ArgoCD, tell me what happens (show me the diff) if this content of this specific branch was applied." You get it?

**Reply by nastacio** [reply]
Would these two options work? (I have not tried them yet)

```
      --local string             Compare live app to a local manifests
      --local-repo-root string   Path to the repository root. Used together with --local allows setting the repository root (default "/")
```

**Reply by rishabh625** [reply]
yeah seems reasonable for this usecase
```
--local string             Compare live app to a local manifests
      --local-repo-root string   Path to the repository root. Used together with --local allows setting the repository root (default "/")
```

**Reply by himanshuupadhyay101** [reply]
@edsonmarquezani Hello, did you find out any way to preview diff with argocd?

**Comment by nrnrk** [comment, marked as answer]
I found the more suitable flag for me, after I checked this issue. You can check the preview with this flag as before.

```
      --revision string             Compare live app to a particular revision
```

If you set `main` branch as target, you can check the diff for each PR by the following command.

```shell
argocd app diff <app-name> --revision <pr-branch>
```

**Reply by edsonmarquezani** [reply]
Oh, thanks, I ended up not watching the issue anymore, but this seems like what I had been looking for. Thank you.

**Comment by act-mreeves** [comment]
This doesn't seem that useful to me. It doesn't calculate the diff in the k8s resources that will result from (for example) a helm chart version change. 

I am using an app of apps pattern but this shows the diff tool looks nothing like what the argocd diff would look like which **does** render all the k8s changes. 

```
$ argocd app diff argocd/bootstrap-kube-system-app-of-apps --revision ACTTO-2193_argo_repo_diff_exp 

===== argoproj.io/Application argocd/cluster-autoscaler ======
197c197
<     targetRevision: 9.32.0
---
>     targetRevision: 9.36.0
```
vs.
```
$ helm diff -C1 upgrade -n kube-system --version 9.36.0 -f values.yaml autoscaler autoscaler/cluster-autoscaler
kube-system, autoscaler-aws-cluster-autoscaler, ClusterRole (rbac.authorization.k8s.io) has changed:
...
      app.kubernetes.io/managed-by: "Helm"
-     app.kubernetes.io/version: "1.27.2"
-     helm.sh/chart: "cluster-autoscaler-9.32.0"
+     app.kubernetes.io/version: "1.29.0"
+     helm.sh/chart: "cluster-autoscaler-9.36.0"
    name: autoscaler-aws-cluster-autoscaler
...
      - list
+     - create
+     - delete
      - get
...
        - watch
+       - get
    - apiGroups:
...
```
